File: 0801-0900/0894-All Possible Full Binary Trees/0894-All Possible Full Binary Trees.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    memo = {0: [], 1: [TreeNode(0)]}
    def allPossibleFBT(self, N: int) -> List[TreeNode]:
        if N not in Solution.memo:
            result = []
            for i in range(1, N, 2):
                lefts = self.allPossibleFBT(i)
                rights = self.allPossibleFBT(N - 1 - i)
                for left in lefts:
                    for right in rights:
                        root = TreeNode(0)
                        root.left = left
                        root.right = right
                        result.append(root)
            Solution.memo[N] = result
        return Solution.memo[N]

File: 0801-0900/0894-All Possible Full Binary Trees/0894-All Possible Full Binary Trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> allPossibleFBT(int N) {
        if (N == 0) {
            return {};
        }
        if (N == 1) {
            return {new TreeNode(0)};
        }
        if (!cache.count(N)) {
            vector<TreeNode*> result;
            for (int i = 1; i < N; i += 2) {
                vector<TreeNode*> lefts = allPossibleFBT(i);
                vector<TreeNode*> rights = allPossibleFBT(N - i - 1);
                for (TreeNode* left : lefts) {
                    for (TreeNode* right : rights) {
                        TreeNode* root = new TreeNode(0);
                        root->left = left;
                        root->right = right;
                        result.push_back(root);
                    }
                }
            }
            cache[N] = result;
        }
        return cache[N];
    }
private:
    unordered_map<int, vector<TreeNode*>> cache;
};
